plugin-localization/src/main/java/com.mapbox.mapboxsdk.plugins.localization/LocalizationPlugin.java

```diff
@@ -0,0 +1,193 @@
+
+import java.util.Locale;
+
+import static com.mapbox.mapboxsdk.style.layers.PropertyFactory.textField;
+
+/**
+ * The localization plugin enables localization of map labels into the user’s preferred language.
+ * Preferred language is determined by the language that the device is set to in the device's settings.
+ * <p>
+ * Initialise this plugin in the {@link com.mapbox.mapboxsdk.maps.OnMapReadyCallback#onMapReady(MapboxMap)} and provide
+ * a valid instance of a {@link MapboxMap}.
+ * </p>
+ */
+public final class LocalizationPlugin {
+
+  private MapboxMap mapboxMap;
+
+  /**
+   * Create a {@link LocalizationPlugin}.
+   *
+   * @param mapboxMap the MapboxMap to apply the localization plugin to
+   * @since 0.1.0
+   */
+  public LocalizationPlugin(@NonNull MapboxMap mapboxMap) {
+    this.mapboxMap = mapboxMap;
+    setMapTextLanguage(getDeviceLanguage());
+  }
+
+  /**
+   * Retrieve the code for the language that the device is currently set to.
+   *
+   * @return The correct code according to ISO-639-1
+   */
+  public String getDeviceLanguage() {
+    if (Build.VERSION.SDK_INT >= 21) {
+      String languageTag = Locale.getDefault().toLanguageTag();
+      if (languageTag.contains("Hans")) {
+        return "zh-Hans";
+      } else if (languageTag.contains("Hant")) {
+        return "zh";
+      } else {
+        return Locale.getDefault().toLanguageTag().substring(0, 2);
+      }
+    } else {
+      if (Locale.getDefault().getLanguage().equalsIgnoreCase("IW")) {
+        return "he";
+      } else if (Locale.getDefault().getLanguage().equalsIgnoreCase("IN")) {
+        return "id";
+      } else if (Locale.getDefault().getLanguage().equalsIgnoreCase("JI")) {
+        return "yi";
+      } else if (Locale.getDefault().getLanguage().equalsIgnoreCase("JW")) {
+        return "jv";
+      } else if (Locale.getDefault().getLanguage().contains("zh-Hans")
+        || Locale.getDefault().getDisplayName().contains("简体")) {
+        return "zh-Hans";
+      } else if (Locale.getDefault().getLanguage().contains("zh-Hant")
+        || Locale.getDefault().getDisplayName().contains("繁體")) {
+        return "zh";
+      } else {
+        return Locale.getDefault().getLanguage();
+      }
+    }
+  }
+
+  /**
+   * Sets map text to the specified language
+   *
+   * @param languageToSetMapTo The language that to set the map text to.
+   * @since 0.1.0
+   */
+  public void setMapTextLanguage(String languageToSetMapTo) {
+    for (Source source : mapboxMap.getSources()) {
+      if (sourceIsFromMapbox(source)) {
+        for (Layer layer : mapboxMap.getLayers()) {
+          if (layerHasAdjustableTextField(layer)) {
+            if (((SymbolLayer) layer).getTextField().getValue().contains("name")) {
+              layer.setProperties(textField(String.format("{name_%s}", languageToSetMapTo)));
+            }
+            if (((SymbolLayer) layer).getTextField().getValue().contains("abbr")
+              && !getDeviceLanguage().equals("en")) {
```

Comment:
This assumes the style was set to use `name_en` in the first place, which is true about most Mapbox-designed styles like Mapbox Streets. However, it’d be totally reasonable for a style to use `name` (local language) right out of the gate.

For reference, the iOS and macOS map SDKs preserve the style’s original language (e.g., English) unless the `localizesLabels` feature is enabled. Once the feature is enabled, it [changes every `name` and `name_*` to the device language](https://github.com/mapbox/mapbox-gl-native/blob/e2a4dffcb627498561d66a304ec668f8242b8342/platform/darwin/src/MGLVectorSource.mm#L118-L122), [falling back to `name_en`](https://github.com/mapbox/mapbox-gl-native/blob/e2a4dffcb627498561d66a304ec668f8242b8342/platform/darwin/src/MGLVectorSource.mm#L99) if the device language is unsupported. So whether the style originally used `name_en`, `name`, or `name_zh`, the localization feature will still take effect.

(The Mapbox GL Language plugin for GL JS has a [`defaultLanguage`](https://github.com/mapbox/mapbox-gl-language/blob/9410fd13ce5ef47b90f91bb9af363557614b360b/index.js#L23) option instead of falling back to English.)

Reply:
Removed the ` && !getDeviceLanguage().equals("en")` assumption in this commit https://github.com/mapbox/mapbox-plugins-android/pull/74/commits/86be093d0f65d60d8dd8c3a86287e7ffe2a0109f 

The comment below addresses the checking and fallback functions that I've built in.